In [59]:
import numpy as np, pandas as pd, matplotlib.pyplot as plt
from scipy.optimize import root_scalar

In [60]:
# 1=N2,2=H2,3=NH3,4=A
# set constants
z=np.array([0.2099,0.6298,0.1050,0.0553])
K=np.array([66.7,50,0.0235,100])


# Define function
def eqns(VoF):
    S=0

    for i in range(4):
        S = S + z[i]*(K[i]-1)/(1 + VoF*(K[i]-1))
        
    f = S
    return f

In [61]:
# Set initial guesses
VoFo1 = 0.5
VoFo2 = 1
# Call fsolve, using secant method require two guesses different from MATLAB
result = root_scalar(eqns, x0 = VoFo1, x1 = VoFo2, method = 'secant') 
VoF = result.root

# Display the zero
print("The root of function is: " ,VoF) 
# Display value of the function at the "zero"
print("The vale of function at the 'zero:'", eqns(VoF))

The root of function is:  0.9145914803976979
The vale of function at the 'zero:' 3.917491331328904e-12


Notice that the root found in python is slight different from the root found in MATLAB. \
Python root: 0.9145914803976979\
MATLAB root: 0.914591480398154\
The last few digit different reuslts in the value of function at the "zero" mismatch.\
Python function value at the "zero":  3.917491331328904e-12\
MATLAB function value at the "zero": -5.443284711859064e-13

In [62]:
x = z/(1 + VoF*(K-1))
y = z*K/(1 + VoF*(K-1))
# L = 1 1bmol as basis
F = 1 / ( 1 - VoF )
V = VoF * F
print("x = ",x,"\ny = ",y,"\nF = ",F,"\nV = ",V)

x =  [3.43598958e-03 1.37465948e-02 9.82213338e-01 6.04077425e-04] 
y =  [0.2291805  0.68732974 0.02308201 0.06040774] 
F =  11.70843382670042 
V =  10.70843382670042
